### ISSUE TITLE
Investigate poor performance of `iterateWordSets` and `wordSetToWords`

### ISSUE THREAD

**Issue by RunDevelopment** [RunDevelopment/refa#31]
On my machine, both methods are pretty slow and I don't quite understand why.

`iterateWordSets` takes about 0.1ms per yield which is quite bad.

`wordSetToWords` takes about 1ms per yield which is really bad.

Task: Find out why these methods are so slow (and maybe try to fix it). Focus on the 10x slowdown of `wordSetToWords`.

**Comment by RunDevelopment** [comment]
52bf8a0caa4842972db339a0613d7c8033208a7a drastically improves the worst-case performance of `wordSetToWords`. It's now up to 50x faster in some cases. The average case is faster also.

`wordSetToWords` should be good enough for now. It's not particularly fast but it should be fast enough. I'm sure that there are still some bad worst-cases (compared to the avg case) but even those will be kinda acceptable performance-wise.

We can now focus on the performance of `iterateWordSets`.

**Comment by RunDevelopment** [comment]
Ok, after some more investigation, I think the performance of `iterateWordSets` is good enough.

"Good enough" as in "it's not unreasonably slow" and as in "I cannot think of any way to make it faster". 

---

I did try out one change. Rn, `iterateWordSets` creates a tree of all words. We can save some memory and time by using a DAG instead, I thought. This sorta worked. It was able to eliminate some worst cases but it made the average case 10x slower. I ultimately decided against this change.

Should the performance issue come up in the future again, the DAG approach is one possible solution. I only did a simple implementation. It might be possible to make it faster. Another idea is to find some heuristic that decides whether the tree or DAG approach should be used.
